### DISCUSSION TITLE
P2G heatmap groupBy cluster

### DISCUSSION THREAD

**Discussion by AnjaliC4** [GreenleafLab/ArchR#1312] [Questions / Documentation]
Hi @rcorces,

I am trying to understand Peak2Gene heatmaps. I see that I can use groupBy to group by clusters, for example. The explanation for this is:
"groupBy | The name of the column in cellColData to use for labeling KNN groupings. The maximum group appeared in the KNN groupings is used."

1. If groupBy= Clusters- how are KNNs of p2g linkages grouped by clusters? What does it mean that "The maximum group appeared in the KNN groupings is used."  Also, I am unable to find the source code for plotPeak2GeneHeatmap - could you kindly point me in the direction?

2. My goal is to get P2G linkages per cluster per condition. Does it make sense to use output of  plotPeak2GeneHeatmap per cluster (from returnMatrices=T) for per-cluster-condition peak-to-gene linkages? What is the best way to obtain this information from ArchR? Thanks

Thanks very much

**Comment by rcorces** [comment]
> how are KNNs of p2g linkages grouped by clusters? What does it mean that "The maximum group appeared in the KNN groupings is used."

It means that all of the k nearest neighbors vote and whichever cluster is represented by the most knns is used for the label of that group

> Also, I am unable to find the source code for plotPeak2GeneHeatmap - could you kindly point me in the direction?

https://github.com/GreenleafLab/ArchR/blob/968e4421ce7187a8ac7ea1cf6077412126876d5f/R/IntegrativeAnalysis.R#L1384

> My goal is to get P2G linkages per cluster per condition

Peak to gene linkages are called based on variance across cells. You need variance to identify linkages. I'm not sure what you mean by "per cluster" but if you are saying what I think you are saying, then this might not work.


Also, referencing the methods of the publication is often a good idea:


> First, ArchR identifies 500+ low-overlapping cell aggregates (see Creation of Low-Overlapping Aggregates of Cells for Linkage Analysis).  Second, ArchR reads in the peak matrix and then creates the cell aggregate x peak matrix. Third, ArchR reads in the gene expression matrix and then creates the cell aggregate x gene matrix. ArchR then identifies all possible peak-to-gene combinations within a given window of the gene start (user-defined, default is 250 kb) and then computes the Pearson correlation of the log2-normalized cell aggregate x peak matrix and cell aggregate x gene matrix across all cell aggregates. ArchR computes these peak-to-gene links genome-wide and stores them within the ArchRProject, which can then be accessed for downstream applications. Additionally, ArchR enables users to lower the resolution of these interactions to better visualize the main interactors (keeping only the highest correlation value observed in each window).

**Reply by alyamahmoud** [reply]
I have a question related to the creation of low overlapping cell aggregates, are these always n = 500 ? also how do these map to any other metadata(proj@peakSet) ? 
I am also confused about the k argument in the plotPeak2GeneHeatmap() which I assumed is going to give 10 low-overlapping cluster aggregated but instead it gives 10 kmeansID (from the p2g$ATAC$kmeansID). 
Thank you

**Comment by so145** [comment]
After obtaining peak-to-gene linkages is it possible in ArchR to take your linked peaks and genes and produce a heatmap of your peak accessibility and gene expression but averaged for each cluster? For example like Fig. 3A in this paper: https://www.nature.com/articles/s41593-021-01002-4#Fig3

**Reply by rcorces** [reply]
We do not provide a function to do this but that is because it is relatively straightforward with standard R bioinformatics. Use `getGroupSE()` to get pseudo-bulked summarizedExperiment objects for the ATAC signal and the gene expression signal. Then organize them to match based on your links. Then plot your heatmaps using `ComplexHeatmap` or a similar package. We cannot provide code to show how to do this.

**Reply by so145** [reply]
Thanks for your help and the quick response.
